File: 351002/Mikhalkov/discussion/Distcomp.Discussion.Infrastructure/Repositories/NoteRepository.cs
using Distcomp.Shared.Models;
using Cassandra;
using Distcomp.Discussion.Infrastructure.Data;

namespace Distcomp.Discussion.Infrastructure.Repositories
{
    public class NoteRepository
    {
        private readonly ISession _session;
        private readonly PreparedStatement _insertStmt;

        public NoteRepository(CassandraProvider provider)
        {
            _session = provider.Session;
            _insertStmt = _session.Prepare("INSERT INTO tbl_note (country, issue_id, id, content, state) VALUES (?, ?, ?, ?, ?)");
        }

        public void Update(Note note) => Save(note);

        public void Save(Note note)
        {
            var statement = _insertStmt.Bind(note.Country, note.IssueId, note.Id, note.Content, note.State.ToString());
            _session.Execute(statement);
        }

        public Note? GetOne(string country, long issueId, long id)
        {
            var row = _session.Execute($"SELECT * FROM tbl_note WHERE country='{country}' AND issue_id={issueId} AND id={id}").FirstOrDefault();
            return row == null ? null : Map(row);
        }

        public bool Exists(string country, long issueId, long id)
        {
            var row = _session.Execute($"SELECT id FROM tbl_note WHERE country='{country}' AND issue_id={issueId} AND id={id}").FirstOrDefault();
            return row != null;
        }

        public IEnumerable<Note> GetAll() =>
            _session.Execute("SELECT * FROM tbl_note").Select(Map);

        public void Delete(string country, long issueId, long id) =>
            _session.Execute($"DELETE FROM tbl_note WHERE country='{country}' AND issue_id={issueId} AND id={id}");

        public Note? GetByIdOnly(long id)
        {
            var row = _session.Execute($"SELECT * FROM tbl_note WHERE id={id} ALLOW FILTERING").FirstOrDefault();
            return row == null ? null : Map(row);
        }

        private Note Map(Row row) => new Note
        {
            Country = row.GetValue<string>("country"),
            IssueId = row.GetValue<long>("issue_id"),
            Id = row.GetValue<long>("id"),
            Content = row.GetValue<string>("content"),
            State = Enum.Parse<NoteState>(row.GetValue<string>("state"))
        };
    }
}

File: 351004/Leshok/lab3/discussion/src/repositories/NoteRepository.js
const { client } = require('../config/cassandra');

class NoteRepository {
  async save(note) {
    const query = `INSERT INTO tbl_note (country, newsId, id, content) VALUES (?, ?, ?, ?)`;
    await client.execute(query, [
      note.country,
      Number(note.newsId),
      Number(note.id),
      note.content
    ], { prepare: true });
    return note;
  }

  async updateFull(note) {
    const query = `UPDATE tbl_note SET content = ? WHERE country = ? AND newsId = ? AND id = ?`;
    await client.execute(query, [
      note.content,
      note.country,
      Number(note.newsId),
      Number(note.id)
    ], { prepare: true });
    return note;
  }

  
  async findByIdOnly(id) {
    const query = `SELECT * FROM tbl_note WHERE id = ? ALLOW FILTERING`;
    const result = await client.execute(query, [Number(id)], { prepare: true });
    if (result.rows.length === 0) return null;
    const row = result.rows[0];
    return {
      id: Number(row.id),
      newsId: Number(row.newsid),
      country: row.country,
      content: row.content
    };
  }

  async findAll() {
    const result = await client.execute(`SELECT * FROM tbl_note`);
    return result.rows.map(row => ({
      id: Number(row.id),
      newsId: Number(row.newsid),
      country: row.country,
      content: row.content
    }));
  }

  async findByNewsId(country, newsId) {
    const query = `SELECT * FROM tbl_note WHERE country = ? AND newsId = ?`;
    const result = await client.execute(query, [country, Number(newsId)], { prepare: true });
    return result.rows.map(row => ({
      id: Number(row.id),
      newsId: Number(row.newsid),
      country: row.country,
      content: row.content
    }));
  }

  async deleteByIdOnly(id) {
    const note = await this.findByIdOnly(id);
    if (!note) return false;
    const query = `DELETE FROM tbl_note WHERE country = ? AND newsId = ? AND id = ?`;
    await client.execute(query, [note.country, note.newsId, Number(id)], { prepare: true });
    return true;
  }

  async updateByIdOnly(id, content) {
    const note = await this.findByIdOnly(id);
    if (!note) return null;
    const query = `UPDATE tbl_note SET content = ? WHERE country = ? AND newsId = ? AND id = ?`;
    await client.execute(query, [content, note.country, note.newsId, Number(id)], { prepare: true });
    return { ...note, content };
  }
}

module.exports = new NoteRepository();